File: opencode/tmux-manager/src/launch.ts
import {
  getGlobalOption,
  displayMessage,
  hasSession,
  newSession,
  setSessionOption,
  displayPopup,
} from './tmux';
import { sessionNameForPath } from './sessions';

export function launch(cwd: string, origin?: string): void {
  const prefix = getGlobalOption('@opencode_session_prefix', 'opencode-');
  const command = getGlobalOption('@opencode_command', 'opencode');
  const width = getGlobalOption('@opencode_popup_width', '90%');
  const height = getGlobalOption('@opencode_popup_height', '90%');

  const currentSession = displayMessage('#S');
  if (currentSession.startsWith(prefix)) {
    displayMessage('Popup window already open');
    return;
  }

  const sessionName = sessionNameForPath(cwd, prefix);

  if (!hasSession(sessionName)) {
    newSession(sessionName, cwd, command);
  }

  setSessionOption(sessionName, '@opencode_path', cwd);
  if (origin) {
    setSessionOption(sessionName, '@opencode_origin', origin);
  }

  displayPopup({
    width,
    height,
    command: `tmux attach-session -t ${sessionName}`,
  });
}

File: llm-session-manager/internal/launch/launch.go
package launch

import (
	"fmt"
	"strings"

	"llm-session-manager/internal/sessions"
	"llm-session-manager/internal/tmux"
)

// Launch creates or reuses a session for cwd and opens it in a popup.
func Launch(cwd, origin string) error {
	prefix := tmux.GetGlobalOption("@llm_session_prefix", "llm-")
	command := tmux.GetGlobalOption("@llm_command", "opencode")
	width := tmux.GetGlobalOption("@llm_popup_width", "90%")
	height := tmux.GetGlobalOption("@llm_popup_height", "90%")

	currentSession, err := tmux.DisplayMessage("#S", "")
	if err != nil {
		// Not running inside tmux; continue with launch.
		currentSession = ""
	}
	if currentSession != "" && currentSession != "__unknown__" && strings.HasPrefix(currentSession, prefix) {
		// Already inside a managed session.
		_, _ = tmux.DisplayMessage("Popup window already open", "")
		return nil
	}

	sessionName := sessions.SessionNameForPath(cwd, prefix)

	if !tmux.HasSession(sessionName) {
		if err := tmux.NewSession(sessionName, cwd, command); err != nil {
			return fmt.Errorf("failed to create session: %w", err)
		}
	}

	if err := tmux.SetSessionOption(sessionName, "@llm_path", cwd); err != nil {
		return err
	}
	if origin != "" {
		if err := tmux.SetSessionOption(sessionName, "@llm_origin", origin); err != nil {
			return err
		}
	}

	return tmux.DisplayPopup(tmux.DisplayPopupOptions{
		Width:   width,
		Height:  height,
		Command: fmt.Sprintf("tmux attach-session -t %s", sessionName),
	})
}
